### ISSUE TITLE
Improve CLI

### ISSUE THREAD

**Issue by bartekpacia** [ohmyzsh/ohmyzsh#9087]
**Is your feature request related to a particular plugin or theme?**
No. This feature would be implemented in [`lib/cli.zsh`](https://github.com/ohmyzsh/ohmyzsh/blob/master/lib/cli.zsh) file.

**Is your feature request related to a problem? Please describe.**
The problem is that e.g changing themes and adding/removing plugins is kind of a time-consuming activity.

**Describe the solution you'd like**
I suggest adding more functionality to the `omz` CLI tool.

Examples:
`omz plugin add <name>` - Adds a new plugin to the `plugins` array in `~/.zshrc`

`omz plugin remove <name>` - Removes a plugin from the `plugins` array in `~/.zshrc`

`omz theme set <name>` - changes the value of the `ZSH_THEME` variable in `~/.zshrc`

That's what I can think of right now.

**Describe alternatives you've considered**
Not doing anything?

**Additional context**
idea by @larson-carter

**Related Issues**
I'm not aware of them, but please let me know if there are any so I can list them here

**Comment by larson-carter** [comment]
This is going to be a massive change. However it is much welcomed and can make omz seem a lot more inclusive.

Don't quote me. However, I believe that there is already an API or tool somewhere that does this functionality. However if I remember correctly it has a few dependencies/device specific requirements that breaks on certain machines. 

So the main implementation would be to make sure that this is a universal tool and not one that requires a specific operating system, build utility, etc.

> The problem is that e.g changing themes and adding/removing plugins is kind of a time-consuming activity.

I think this is a true statement in some situations. After a little bit of tweaking it can become repetitive/annoying so this is a great idea to improve/fix in omz.

> I suggest adding more functionality to the omz CLI tool.
> No. This feature would be implemented in lib/cli.zsh file.

I think that including it in the `lib/cli.zsh` file is the proper call. Since that is the CLI that we use to test PRs and such. I think that it would be unnecessary to add this functionality in another file. I'd like @mcornella's opinion on that too.

> the CLI might also ask whether to reload the shell [Y/n] or use a default option from somewhere (probably ~/.zshrc)

I think that would be necessary. We could have it run another zsh session to reload the shell. So at the end of each execution it would reload the shell.

I can't think of any other issues/problems that would result from some of these changes. I'd like to hear others input.

**Comment by mcornella** [comment]
These are both awesome ideas but it's not an easy feat. The `ZSH_THEME` line is easier to change, but the `plugins` array might span multiple lines. But it is in the making, you won't have to wait much more for a testable prototype.

**Comment by larson-carter** [comment]
> These are both awesome ideas but it's not an easy feat

I agree it's an awesome idea. However it is going to be a pain to implement.

>  the plugins array might span multiple lines

That is something that I was afraid of. Maybe we could implement something like `\n`? To break the line and go down to the next line?

> But it is in the making, you won't have to wait much more for a testable prototype.

Can you link to where this is being done? I can't remember if I've seen it yet.

**Comment by mcornella** [comment]
> Can you link to where this is being done? I can't remember if I've seen it yet.

https://github.com/mcornella/ohmyzsh/tree/feat/ohmyzsh-cli-full

**Comment by R0LA1mRifcF8yAkk** [comment]
Any update on this? Would really love to use the plugin list feature already!

**Comment by larson-carter** [comment]
I currently don't have an official update on this @mcornella will be the person to ask.

I've been messing around with ways to get it to append the `zshrc` file however I have been having weird issues/bugs.

Please let me know @R0LA1mRifcF8yAkk if you have any ideas that would possibly work!

**Comment by R0LA1mRifcF8yAkk** [comment]
@larson-carter I'm sorry, I have little to no experience with shell scripting.

I'm just interested in seeing a list of currently installed zsh plugins because I have a script that lists all development software (and brew casks) that I have installed.

**Comment by mcornella** [comment]
I pushed 1bb402190d53c118f47a0851f5714dd4faa52809 which adds a way to list plugins and themes, and optionally change the theme of the current zsh session (it doesn't persist it into the zshrc file).

The other stuff (changing the zshrc file) will take time to get right.

**Comment by R0LA1mRifcF8yAkk** [comment]
@mcornella  The output of `omz plugin list` is broken for me:
```
[4m[1m[4mCustom plugins[0m[4m[24m: autoupdate example fz you-should-use zsh-autosuggestions zsh-syntax-highlighting

[4m[1m[4mBuilt-in plugins[0m[4m[24m: ...
```

**Comment by mcornella** [comment]
I assume this
```zsh
print -Pn "%U%BCustom plugins%b%u: "
```
also looks broken? If so, can you post the output of `echo $TERM` and `setopt`?

**Comment by R0LA1mRifcF8yAkk** [comment]
`print -P "%U%BCustom plugins%b%u: "` looks fine. Bold and underlined.

#### `echo $TERM`:
`screen-256color`

#### `setopt`:

- alwaystoend
- autocd
- autopushd
- combiningchars
- completeinword
- correctall
- extendedhistory
- noflowcontrol
- histexpiredupsfirst
- histignoredups
- histignorespace
- histverify
- interactive
- interactivecomments
- longlistjobs
- monitor
- promptsubst
- pushdignoredups
- pushdminus
- sharehistory
- shinstdin
- zle

**Comment by mcornella** [comment]
Hmm can you run `typeset -f -t _omz::plugin::list` and run `omz plugin list`, and post the output?

**Comment by R0LA1mRifcF8yAkk** [comment]
```
+_omz::plugin::list:1> local -a custom_plugins builtin_plugins
+_omz::plugin::list:2> custom_plugins=( autoupdate example fz you-should-use zsh-autosuggestions zsh-syntax-highlighting )
+_omz::plugin::list:3> builtin_plugins=( adb alias-finder ansible ant apache2-macports arcanist archlinux asdf autoenv autojump autopep8 aws battery bazel bbedit bgnotify boot2docker bower branch brew bundler cabal cake cakephp3 capistrano cargo cask catimg celery chruby chucknorris cloudapp cloudfoundry codeclimate coffee colemak colored-man-pages colorize command-not-found common-aliases compleat composer copybuffer copydir copyfile cp cpanm dash debian dircycle direnv dirhistory dirpersist django dnf dnote docker docker-compose docker-machine doctl dotenv dotnet droplr drush eecms emacs ember-cli emoji emoji-clock emotty encode64 extract fabric fancy-ctrl-z fasd fastfile fbterm fd fedora firewalld flutter forklift fossil frontend-search fzf gas gatsby gb gcloud geeknote gem git git-auto-fetch git-escape-magic git-extras git-flow git-flow-avh git-hubflow git-lfs git-prompt gitfast github gitignore glassfish globalias gnu-utils go golang gpg-agent gradle grails grunt gulp hanami helm heroku history history-substring-search hitokoto homestead httpie ionic ipfs iterm2 jake-node jenv jfrog jhbuild jira jruby jsontools jump kate keychain kitchen knife knife_ssh kops kube-ps1 kubectl laravel laravel4 laravel5 last-working-dir lein lighthouse lol lxd macports magic-enter man marked2 mercurial meteor microk8s minikube mix mix-fast mosh mvn mysql-macports n98-magerun nanoc ng nmap node nomad npm npx nvm oc osx otp pass paver pep8 per-directory-history percol perl perms phing pip pipenv pj please pod postgres pow powder powify profiles pyenv pylint python rails rake rake-fast rand-quote rbenv rbfu react-native rebar redis-cli repo ripgrep ros rsync ruby rust rustup rvm safe-paste salt sbt scala scd screen scw sdk sfdx sfffe shell-proxy shrink-path singlechar spring sprunge ssh-agent stack sublime sublime-merge sudo supervisor suse svcat svn svn-fast-info swiftpm symfony symfony2 systemadmin systemd taskwarrior terminitor terraform textastic textmate thefuck themes thor tig timer tmux tmux-cssh tmuxinator torrent transfer tugboat ubuntu ufw urltools vagrant vagrant-prompt vault vi-mode vim-interaction virtualenv virtualenvwrapper vscode vundle wakeonlan wd web-search wp-cli xcode yarn yii yii2 yum z zeus zsh-interactive-cd zsh-navigation-tools zsh_reload )
+_omz::plugin::list:5> ((  6  ))
+_omz::plugin::list:6> print -Pn '%U%BCustom plugins%b%u: '
+_omz::plugin::list:7> print -l autoupdate example fz you-should-use zsh-autosuggestions zsh-syntax-highlighting
+_omz::plugin::list:8> fmt -w 238
[4m[1m[4mCustom plugins[0m[4m[24m: autoupdate example fz you-should-use zsh-autosuggestions zsh-syntax-highlighting
+_omz::plugin::list:10> ((  283  ))
+_omz::plugin::list:12> ((  6  ))
+_omz::plugin::list:12> echo
+_omz::plugin::list:14> print -Pn '%U%BBuilt-in plugins%b%u: '
+_omz::plugin::list:16> fmt -w 238
+_omz::plugin::list:15> print -l adb alias-finder ansible ant apache2-macports arcanist archlinux asdf autoenv autojump autopep8 aws battery bazel bbedit bgnotify boot2docker bower branch brew bundler cabal cake cakephp3 capistrano cargo cask catimg celery chruby chucknorris cloudapp cloudfoundry codeclimate coffee colemak colored-man-pages colorize command-not-found common-aliases compleat composer copybuffer copydir copyfile cp cpanm dash debian dircycle direnv dirhistory dirpersist django dnf dnote docker docker-compose docker-machine doctl dotenv dotnet droplr drush eecms emacs ember-cli emoji emoji-clock emotty encode64 extract fabric fancy-ctrl-z fasd fastfile fbterm fd fedora firewalld flutter forklift fossil frontend-search fzf gas gatsby gb gcloud geeknote gem git git-auto-fetch git-escape-magic git-extras git-flow git-flow-avh git-hubflow git-lfs git-prompt gitfast github gitignore glassfish globalias gnu-utils go golang gpg-agent gradle grails grunt gulp hanami helm heroku history history-substring-search hitokoto homestead httpie ionic ipfs iterm2 jake-node jenv jfrog jhbuild jira jruby jsontools jump kate keychain kitchen knife knife_ssh kops kube-ps1 kubectl laravel laravel4 laravel5 last-working-dir lein lighthouse lol lxd macports magic-enter man marked2 mercurial meteor microk8s minikube mix mix-fast mosh mvn mysql-macports n98-magerun nanoc ng nmap node nomad npm npx nvm oc osx otp pass paver pep8 per-directory-history percol perl perms phing pip pipenv pj please pod postgres pow powder powify profiles pyenv pylint python rails rake rake-fast rand-quote rbenv rbfu react-native rebar redis-cli repo ripgrep ros rsync ruby rust rustup rvm safe-paste salt sbt scala scd screen scw sdk sfdx sfffe shell-proxy shrink-path singlechar spring sprunge ssh-agent stack sublime sublime-merge sudo supervisor suse svcat svn svn-fast-info swiftpm symfony symfony2 systemadmin systemd taskwarrior terminitor terraform textastic textmate thefuck themes thor tig timer tmux tmux-cssh tmuxinator torrent transfer tugboat ubuntu ufw urltools vagrant vagrant-prompt vault vi-mode vim-interaction virtualenv virtualenvwrapper vscode vundle wakeonlan wd web-search wp-cli xcode yarn yii yii2 yum z zeus zsh-interactive-cd zsh-navigation-tools zsh_reload

[4m[1m[4mBuilt-in plugins[0m[4m[24m: adb alias-finder ansible ant apache2-macports arcanist archlinux asdf autoenv autojump autopep8 aws battery bazel bbedit bgnotify boot2docker bower branch brew bundler cabal cake cakephp3 capistrano
cargo cask catimg celery chruby chucknorris cloudapp cloudfoundry codeclimate coffee colemak colored-man-pages colorize command-not-found common-aliases compleat composer copybuffer copydir copyfile cp cpanm dash debian dircycle direnv
dirhistory dirpersist django dnf dnote docker docker-compose docker-machine doctl dotenv dotnet droplr drush eecms emacs ember-cli emoji emoji-clock emotty encode64 extract fabric fancy-ctrl-z fasd fastfile fbterm fd fedora firewalld
flutter forklift fossil frontend-search fzf gas gatsby gb gcloud geeknote gem git git-auto-fetch git-escape-magic git-extras git-flow git-flow-avh git-hubflow git-lfs git-prompt gitfast github gitignore glassfish globalias gnu-utils go
golang gpg-agent gradle grails grunt gulp hanami helm heroku history history-substring-search hitokoto homestead httpie ionic ipfs iterm2 jake-node jenv jfrog jhbuild jira jruby jsontools jump kate keychain kitchen knife knife_ssh kops
kube-ps1 kubectl laravel laravel4 laravel5 last-working-dir lein lighthouse lol lxd macports magic-enter man marked2 mercurial meteor microk8s minikube mix mix-fast mosh mvn mysql-macports n98-magerun nanoc ng nmap node nomad npm npx nvm
oc osx otp pass paver pep8 per-directory-history percol perl perms phing pip pipenv pj please pod postgres pow powder powify profiles pyenv pylint python rails rake rake-fast rand-quote rbenv rbfu react-native rebar redis-cli repo ripgrep
ros rsync ruby rust rustup rvm safe-paste salt sbt scala scd screen scw sdk sfdx sfffe shell-proxy shrink-path singlechar spring sprunge ssh-agent stack sublime sublime-merge sudo supervisor suse svcat svn svn-fast-info swiftpm symfony
symfony2 systemadmin systemd taskwarrior terminitor terraform textastic textmate thefuck themes thor tig timer tmux tmux-cssh tmuxinator torrent transfer tugboat ubuntu ufw urltools vagrant vagrant-prompt vault vi-mode vim-interaction
virtualenv virtualenvwrapper vscode vundle wakeonlan wd web-search wp-cli xcode yarn yii yii2 yum z zeus zsh-interactive-cd zsh-navigation-tools zsh_reload
```

**Comment by mcornella** [comment]
Looks like `fmt` is doing something weird. Which `fmt --version` are you using?

**Comment by R0LA1mRifcF8yAkk** [comment]
i am not familiar with fmt, output is as follows, even after `brew install fmt`:
```
$ fmt --version

fmt: illegal option -- -
usage:   fmt [-cmps] [-d chars] [-l num] [-t num]
             [-w width | -width | goal [maximum]] [file ...]
Options: -c     center each line instead of formatting
         -d <chars> double-space after <chars> at line end
         -l <n> turn each <n> spaces at start of line into a tab
         -m     try to make sure mail header lines stay separate
         -n     format lines beginning with a dot
         -p     allow indented paragraphs
         -s     coalesce whitespace inside lines
         -t <n> have tabs every <n> columns
         -w <n> set maximum width to <n>
         goal   set target width to goal

```

**Comment by mcornella** [comment]
Ok and `print -P "%U%BCustom plugins%b%u: " | fmt` definitely breaks right?

**Comment by R0LA1mRifcF8yAkk** [comment]
Yeap: `[4m[1m[4mCustom plugins[0m[4m[24m:`

**Comment by mcornella** [comment]
What about `printf "\e[4m\e[1mCustom themes\e[0m: " | fmt`?

**Comment by R0LA1mRifcF8yAkk** [comment]
`[4m[1mCustom themes[0m:`

**Comment by mcornella** [comment]
What about `omz theme list | sed 's/\(\x1b?\[[0-9]*m\)/\x1b\1/g'`?

**Comment by R0LA1mRifcF8yAkk** [comment]
```
[4m[1m[4mCustom themes[0m[4m[24m: example powerlevel10k/powerlevel10k powerlevel10k/powerlevel9k

[4m[1m[4mBuilt-in themes[0m[4m[24m: 3den Soliah adben af-magic afowler agnoster alanpeabody amuse apple arrow aussiegeek avit awesomepanda bira blinks bureau candy candy-kingdom clean cloud crcandy crunch cypher dallas darkblood
daveverwer dieter dogenpunk dpoggi dst dstufft duellj eastwood edvardm emotty essembeh evan fino fino-time fishy flazz fletcherm fox frisk frontcube funky fwalch gallifrey gallois garyblessington gentoo geoffgarside gianu gnzh gozilla
half-life humza imajes intheloop itchy jaischeema jbergantine jispwoso jnrowe jonathan josh jreese jtriley juanghurtado junkfood kafeitu kardan kennethreitz kiwi kolo kphoen lambda linuxonly lukerandall macovsky maran mgutz mh
michelebologna mikeh miloshadzic minimal mira mlh mortalscumbag mrtazz murilasso muse nanotech nebirhos nicoulaj norm obraun peepcode philips pmcgee pygmalion pygmalion-virtualenv random re5et refined rgm risto rixius rkj rkj-repos
robbyrussell sammy simonoff simple skaro smt sonicradish sorin sporty_256 steeef strug sunaku sunrise superjarin suvash takashiyoshida terminalparty theunraveler tjkirch tjkirch_mod tonotdo trapd00r wedisagree wezm wezm+ wuffers
xiong-chiamiov xiong-chiamiov-plus ys zhann
```

this will soon start to feel like trolling

**Comment by mcornella** [comment]
Ok one last time: `omz theme list | sed -E 's/\x1b?(\[[0-9]*m)/\x1b\1/g'`. Sorry I don't have a Mac so I can't test it any other way.

**Comment by mcornella** [comment]
Nvm, this should work (tested on GitHub Actions macOS runner): `omz theme list | sed -E 's/'$'\x1b''?(\[[0-9]*m)/'$'\x1b''\1/g'`. It's ugly, but it should work.

**Comment by R0LA1mRifcF8yAkk** [comment]
Yup, that works, thank you! Any way to only output every _installed_ plugin?

**Comment by mcornella** [comment]
If you mean enabled, that's just running `echo $plugins`.

**Comment by mcornella** [comment]
> Yup, that works, thank you! Any way to only output every _installed_ plugin?

Fixed in 982d010be5fdb1cff471978634f8ff374deba9ba.

**Comment by R0LA1mRifcF8yAkk** [comment]
Now it works as intended. Although I noticed `omz help` doesn't list `plugin` as an option.

**Comment by mcornella** [comment]
Fixed, thanks!

**Comment by carlocab** [comment]
Hi @mcornella (and other maintainers), thanks for the work you put into omz.

I just wanted to mention, just in case, that `omz plugin list` in the master branch will not include plugins that are included in `$ZSH_CUSTOM/plugins` as symlinks.

One instance where you might want to do this is if you have a package installed by your package manager that ships with a zsh plugin but is installed elsewhere. For example, with a standard Homebrew installation, one can find the zsh plugin for z.lua in `/usr/local/opt/z.lua/share/z.lua`. I've put a symlink to this directory in `$ZSH_CUSTOM/plugins`, but `omz plugin list` checks `$ZSH_CUSTOM/plugins` only for directories.

Not sure if you want to fix this, but I thought it might be useful to point this out here if you do.

**Comment by larson-carter** [comment]
@carlocab Great find!

Can you possibly simlink the folder?

**Comment by carlocab** [comment]
Hey @larson-carter, I'm afraid I don't understand the question. I did symlink the folder containing the plugin to `$ZSH_CUSTOM/plugins`, and added the appropriate plugin to the `plugins` array.

The plugin works perfectly fine (no surprises there), but it is not listed by `omz plugin list`. This is because `omz plugin list` checks `$ZSH_CUSTOM/plugins` for directories but ignores symlinks (perhaps because they look just like regular files according to `test -d`).

**Comment by carlocab** [comment]
Perhaps a screenshot would be more informative. `z.lua` is in a slightly different shade because it's a symlink to somewhere else, so `omz plugin list` does not include it as an available plugin.
<img width="1029" alt="Screenshot 2020-10-20 at 01 22 08" src="https://user-images.githubusercontent.com/30379873/96525467-e9062d80-1272-11eb-8b01-a0a61a95a48e.png">

**Comment by larson-carter** [comment]
@carlocab Sorry, I should have made myself more clear, however your screenshot helped out.

Can you run `sudo tree -a` in your custom plugins directory then attach a screenshot? I might have a theory of why It's not working.

**Comment by carlocab** [comment]
@larson-carter no worries; I think I was also rather unclear the first two times too.

I did `sudo exa -a --tree` (I don't have `tree` installed, so I hope this suffices. I can install the version on Homebrew if necessary), and it just shows the contents of each of the plugin directories. There are a lot of files and subdirectories in there, but here's a screenshot of what I suspect you are looking for:

<img width="1029" alt="Screenshot 2020-10-20 at 01 36 41" src="https://user-images.githubusercontent.com/30379873/96526295-105dfa00-1275-11eb-81ee-3a0fbd65329c.png">

Let me know if you need the rest of the output of `sudo exa -a --tree`, or if you actually need the output from `tree` specifically.

**Comment by larson-carter** [comment]
@carlocab 

This screenshot is perfect! I wonder if you can symlink a folder? For example `custom/plugins/z.lua/z.lua.plugin/`

Does that make sense?? I think it's trying to read inside of a folder instead of just the plugin itself.

So basically lets say I download `test.plugin` and it is in my downloads folder at `/home/larsoncarter/Downloads/test.plugin/` I would symlink it to `~/.oh-my-zsh/custom/plugins/test/test.plugin` hence the name I would add to my .zshrc config would be `test`

`plugins(test)`

and hopefully it would also show up when running `omz plugin list`

This is just a theory and I do not know if it will work since I have not yet tested it myself.

**Comment by carlocab** [comment]
So I'm a little confused by your explanation (sorry!), but I think you're suggesting the (obvious in retrospect) workaround of just creating a `z.lua` folder in `$ZSH_CUSTOM/plugins`, and then symlinking `z.lua.plugin.zsh` into the`$ZSH_CUSTOM/plugins/z.lua` folder that I created. Did I understand you correctly?

I guess the way that glob qualifiers work in zsh is that it doesn't think symlinks to directories are directories. From what I can tell from the code, `omz plugin list` finds the custom plugins using

    custom_plugins=("$ZSH_CUSTOM"/plugins/*(/N:t))

The glob qualifier `/` tells zsh to expand `*` to include only directories, and it, unfortunately, does not consider symlinks to directories as directories. I tested this by running the command I show above in the shell:

<img width="795" alt="Screenshot 2020-10-20 at 01 51 26" src="https://user-images.githubusercontent.com/30379873/96526956-e0175b00-1276-11eb-8815-39869836c4af.png">

However, I guess the solution is to not symlink the folder, but just create a folder that contains a symlink.

**Comment by mcornella** [comment]
This is fixable in the glob by adding `-` to the modifiers, which makes zsh follow the symlink target. I pushed f9f45ca2 which hopefully fixes it.

**Comment by carlocab** [comment]
@mcornella Working now! Had to restart my terminal first though. Thanks for looking into this.

**Comment by larson-carter** [comment]
@carlocab Yep @mcornella's commit that he pushed fixed the problem. Basically making it smarter when reading in the directories.

**Comment by mcornella** [comment]
Update: #9872 added the `omz plugin load` subcommand for loading plugins for the current zsh session (a zsh restart will lose the changes, for that you should track #9869).

**Comment by mcornella** [comment]
Update: #9869 added `omz plugin enable` and `omz plugin disable`. It uses `awk` so the generated .zshrc file might not be totally clean, but please test it and open new bug reports if you find any issues.

The last remaining feature request from this issue is `omz theme set`, which will be implemented soon.

**Comment by mcornella** [comment]
Last update: https://github.com/ohmyzsh/ohmyzsh/commit/cbb534267aca09fd123635fc39a7d00c0e21a5f7 introduced the `omz theme set` command. These are still experimental so if you have an issue, the command moves the previous .zshrc file to `.zshrc.bck`, which you can go back to.

Please open a bug report if you find any issues.
